task_2/main.cpp

```diff
@@ -0,0 +1,197 @@
+#include <iostream>
+#include <typeinfo>
+#include <exception>
+
+using std::cout;
+using std::endl;
+using std::string;
+
+struct ICloneable {
+	virtual ICloneable* clone() const = 0;
+	virtual ICloneable* move() = 0;
+	virtual ~ICloneable() { }
+};
+
+template <typename T>
+struct ValueHolder : ICloneable {
+	T data_;		
+	ValueHolder(const T & arg) : data_ {arg} {}
+	ValueHolder * clone() const { 
+		return new ValueHolder(this->data_); 
+	}
+	ValueHolder * move() { 
+		return this; 
+	}
+};
+
+class Any {
+	ICloneable * ptr;			// pointer ICloneable * of base class can contain any type of it template child:
+						// ValueHolder<bool>, ValueHolder<int> or ValueHolder<double>
+public:
+	Any() : ptr {0} {}
+
+	template <typename T>
+	Any(const T & arg) {
+		string type = typeid(T).name();
+		if (type != "i" || type != "d" || type != "b")
+			ptr = 0;
+		ptr = new  ValueHolder<T>(arg);
+	}
+
+	~Any() { delete ptr; };
+
+	Any(const Any & arg){
+		if (arg.ptr == 0)
+			ptr = 0;
+		else
+			ptr = arg.ptr->clone();
+	}
+
+	Any& operator=(Any const & arg) {						
+		if (this != & arg){
+			delete ptr;
+			ptr = arg.ptr->clone();
+		}
+		return * this;
+	}
+	
+	Any(Any && arg){
+		if (arg.ptr == 0)
+			ptr = 0;
+		else {		
+			ptr = arg.ptr->move();	
+			arg.ptr = 0;
+		}
+	}
+
+	Any& operator=(Any && arg) {						
+		if (this != & arg){
+			if (ptr)	
+				delete ptr;
+			ptr = arg.ptr->move();
+			arg.ptr = 0;
+		}
+		return * this;
+	}
+
+	template <typename T>
+	T * cast() const {
+		ValueHolder<T> * tmp = dynamic_cast<ValueHolder<T> *>(ptr);
+		if (tmp == 0)
+			return 0;
+		return &(tmp->data_);
+	}
+
+	template <typename T>
+	T castTo() const {		
+		ValueHolder<T> * tmp = dynamic_cast<ValueHolder<T> *>(ptr);
+		if (tmp == 0)
+			throw Any::BadCastTo();
+		return tmp->data_;
+	}
+	
+	string type() const {
+		string type = typeid(*ptr).name();
+		if (type == "11ValueHolderIiE") return "int";
+		if (type == "11ValueHolderIbE") return "bool";
+		if (type == "11ValueHolderIdE") return "double";
+		return "empty";
+	}
+	static void swap(Any & lhs, Any & rhs){
+		Any tmp = lhs;
+		lhs = rhs;
+		rhs = tmp;
```

Comment:
It is better to use std::move() here to show that you are moving values instead of copying them.

Reply:
ok